proposals/21-11-transformer-annotations.md

```diff
@@ -0,0 +1,253 @@
+
+```yaml
+config.kubernetes.io/transformations: |
+  - kind: PrefixSuffixTransformer
+    path: ../overlays/kustomization.yaml
+    type: builtin
+  - kind: InflateHelmChart
+    path: inflatehelmchart.yaml
+    type: custom
+  - kind: NamespaceTransformer
+    path: examples/multibases/base/kustomization.yaml
+    repo: https://github.com/kubernetes-sigs/kustomize
+    ref: v1.0.6 
+    type: builtin
+```
+
+### Kustomize edit
+
+We will want to provide convenient commands for users to edit the `buildMetadata` field in their kustomization.
+We should support the following commands:
+
+`kustomize edit add buildMetadata originAnnotations`
+
+`kustomize edit add buildMetadata transformerAnnotations`
+
+`kustomize edit add buildMetadata *`
+
+`kustomize edit remove buildMetadata originAnnotaitons`
+
+`kustomize edit remove buildMetadata transformerAnnotations`
+
+`kustomize edit remove buildMetadata *`
+
+The wildcard match `*` should add/remove all possible options for `buildMetadata`.
+ 
+
+### User Stories
+
+#### Story 1
+
+I am running kustomize build on some configuration that has already been created. This configuration has a large number
+of resources and patches through multiple overlays in different directories. After running `kustomize build`, I get a 
+stream of 100+ resources, but I notice that a few of them have a typo, likely originating from the same base or patch 
+file. However, because my directory structure is large and complex, it is difficult to narrow down which base these
+resources originated from and which patches have been applied to it. In the top-level kustomization file, I can add a
+new `buildMetadata` field:
+
+```yaml
+buildMetadata: [originAnnotations, transformerAnnotations]
+```
+
+The output of `kustomize build` will contain patched resources with annotations similar to the following:
+```
+annotations:
+  config.kubernetes.io/origin: |
+    path: deployment.yaml
+  config.kubernetes.io/transformations:
+  - kind: PatchTransformer
+    path: ../base/kustomization.yaml
+    type: builtin
+  - kind: PatchStrategicMergeTransformer
+    path: ../dev/patch.yaml
+    type: builtin
+  - kind: PatchesJson6902Transformer
+    path: patch.yaml
+    type: builtin
+```
+
+From these annotations, I can narrow down the set of files that contributed to the resource's final output and
+debug it accordingly. 
+
+### Risks and Mitigations
+N/A
+
+### Dependencies
+N/A
+
+### Scalability
+N/A
+
+## Drawbacks
+N/A
+
+## Alternatives
+We considered putting this data in the comments of the rendered resources, but this would be a large project because 
+kustomize doesn't currently support comments. 
+
+We also considered making this a flag to `kustomize build` rather than additional field, but we would like to align 
+with the kustomize way of avoiding build-time side effects and have everything declared explicitly in the kustomization.
```

Comment:
Just to extend the motivations of this feature a little bit:
1. This feature is an extension of the previous `originAnnotations`. It may make less sense to split the feature to a separate command and not provide an equivalence declarative approach. 
2. The `kustomize cfg` approach is harder to scale for a large set of resources, which is how the debugging can mostly help with. For example, when it's hard to triage the origin and transformer of 500 deployments from 1000 resource files, users would be unwillingly to run the `kustomize cfg` cmd 500 times I guess?
3. "annotations" is designed to attach metadata, no size-limitations. It is ideal for this kind of tracing info and many other tools or libs use it the same way. e.g. Enterprise level: Google Config Sync, OSS: tekton, skaffold. 
4. When thinking about this approach in the bigger picture, say working in a gitOps CICD approach, the git diff of different kustomize output will not only shows the config change, but the buildMetadata origin changes. It will greatly benefit those users who apply kustomize in their enterprise CICD workflow.

Reply:
+1 to all points Yuwen described above. While I think something like `kustomize cfg explain . --resource=Deployment/default/foo` could be useful in some scenarios, I particularly agree with Yuwen's points 2 and 3 above - scalability for a user needing to inspect a large set of resources, and that this metadata is useful for other tools and "annotations" is where they'd expect to find them.

Reply:
While the proposed implementation clearly builds on originAnnotations, the primary use case set out sounds pretty different, so I think it's worth considering what the best way to achieve the end user goal in question would be rather than presupposing the implementaiton.

In case (2), what is the ideal user workflow in your opinion? Surely they aren't actually looking through all 500 deployments, but rather targeting one or two that have the problem they're investigating. If there's a use case where they really need to delve through all of the large dataset, let's add it.

Re (3), storage isn't actually free, no matter what API server validations will allow. I'm aware of cases of extremely verbose annotations causing significant scalability problems. So if we're designing this feature with the intent that end users will be committing and deploying the data we write, I think we should indeed be careful about how verbose we're being. On the contrary, if the 99% use case is to debug something before commit, we should be super verbose and as helpful as possible.

Further to that point: if the purpose is debugging, it would be most useful to use absolute paths, not paths relative to the parent Kustomization or the current Kustomize root. If we're writing something to be committed, we obviously cannot do that, and in fact should aim to make the annotation content as stable as possible.

Can you give a more detailed example of why (4) is desirable? By embedding what are effectively implementation details of the generation pipeline in the output, we will make it inherently less stable--not generally a good thing for gitOps. For example, renaming a file will cause a diff even though it does not affect the resources' content. Back to my point above, I think optimizing for content that will be committed is in tension with optimizing for debuggability.

Reply:
To add on to Yuwen's point 3, one of the use cases is Google's Config Sync, which takes a kustomization repo, automatically runs `kustomize build`, and syncs it to a cluster. For these users, it is useful to have these annotations available in the cluster for all deployed resources. If a user notices that something is wrong in the cluster, Config Sync can quickly and easily show the user how the resource was created. An open source tool with a very similar requirement is skaffold, which can run kustomize and deploy the output to a cluster via `skaffold deploy`. `kubectl apply -k` also falls under this category.

For a user using kustomize CLI directly, a separate command to debug the output of kustomize makes sense. But for users using other tools that automatically run kustomize and deploy, it is better for them if the annotations persist to the cluster. Many of these users may not even have the kustomize CLI installed to their local machines, and so allowing other tools to add these annotations as part of `kustomize build` is a huge benefit.

Reply:
Thanks Natasha. Yeah, the actual use case is based on the fact that kustomize is integrated into with some other automated tools and users do not directly interact with kustomize.

Reply:
I have updated the proposal to include a user story as described for our use cases, and added adding a separate kustomize command as a considered alternative. Let me know if you feel strongly about having both the command and the built option; otherwise I believe all of the feedback has been addressed and incorporated into the proposal.